### ISSUE TITLE
constants generated by a block are not flagged by MutableConstant cop

### ISSUE THREAD

**Issue by betesh** [bbatsov/rubocop#4150]
Given the following example:

```
$ cat example.rb 
# frozen_string_literal: true

def to_f(i)
  i.to_f
end

CONST1 = [1, 2, 3]

CONST2 = CONST1.map do |i|
  to_f(i)
end

CONST3 = CONST1.map { |i| to_f(i) }
```

the MutableConstant cop should identify violations in `CONST1`, `CONST2` and `CONST3`, but it only identifies `CONST1`.

I was suspicious that this is by design, but I looked around and couldn't find any discussion in the docs or in issues and pull requests indicating that this was ever discussed before.

--------

## Expected behavior

```
1 file inspected, no offenses detected
```

## Actual behavior

```
$ rubocop -- example.rb 
Inspecting 1 file
C

Offenses:

example.rb:7:10: C: Style/MutableConstant: Freeze mutable objects assigned to constants.
CONST1 = [1, 2, 3]
         ^^^^^^^^^

1 file inspected, 1 offense detected
```

## Steps to reproduce the problem

Paste the example above into a file.  Run rubocop on the file.

## RuboCop version

```
$ rubocop -V
0.47.1 (using Parser 2.4.0.0, running on ruby 2.3.0 x86_64-linux)
```

**Comment by betesh** [comment]
One thing to keep in mind is that if we do address this, we need to make sure we don't create an infinite loop when attempting to autocorrect this and `Style/MethodCalledOnDoEndBlock`

**Comment by bbatsov** [comment]
The problem is that you can never be certain whether some method returned something mutable or not. IMO this can't really be improved beyond, at least - not by much.

**Comment by betesh** [comment]
If I understand your correctly, if we just assumed all SCREAMING_SNAKE_CASE variable assignments that were computed by calling a method were mutable, we'd end up with some false positives, i.e. redundant calls to `freeze`.

IMO this would still be an improvement.  I'd rather have a redundant call to `freeze` (which I can explicitly deactive using `# rubocop:disable` if it really bothers me) that forget a call to freeze that should be there.

**Comment by bbatsov** [comment]
Perhaps... I don't have a strong opinion on the subject and I don't like think that there's any good solution to this problem. I don't like solutions that produce false positives in general, even if they are relatively safe like here.

**Comment by jonas054** [comment]
I agree. We usually try as hard as we can to avoid false positives. Closing.
